### ISSUE TITLE
Update log handling to ensure metrics are calculated correctly across versions

### ISSUE THREAD

**Issue by NateWr** [pkp/pkp-lib#4904]
The code which reads access logs and stores metrics data needs to be updated to ensure stats are calculated correctly across different versions.

The main URLs for a submission should stay the same. However, new URLs will be introduced for each version and its galleys (see #4870). Visits to these URLs should go toward a submission's total.

Also, COUNTER may have some rules regarding counting duplicate visits within a time period to the same resource. We need to figure out what these rules specify and how to correctly count visits to two versions of the same item in a short period.

**Comment by NateWr** [comment]
@ctgraham if you have any thoughts on this, that'd be much appreciated!

**Comment by ctgraham** [comment]
This may be a bit broad for this ticket, but I think it is strongly relevant....

James and I chatted with Paul Needham of IRUS-UK last month about some of the considerations in log processing.  This includes:

- Processing logfiles locally vs. [transmitting raw logs](https://irus.jisc.ac.uk/documents/TrackerProtocol-V3-2017-03-22.pdf) for external processing
- Processing bot traffic for exclusion
  - via User Agent
  - via IP
  - via crawl pattern
- Internal datastructures for representing usage at scale
  - We use a flat table amalgamating multiple statistical dimensions
  - IRUS-UK uses tables: suppliers, items, and statistics, with statistics represented by item-month.
- Increased interdependency in R5 of COUNTER-SUSHI as a unified webservice.

There might be opportunities to collaborate with IRUS-UK to architect a common library which can process logs into statistics (or perhaps just perform the bot-exclusions).  It also made we wonder why there wouldn't be a shared library for representing statistics ontop of COUNTER-SUSHI R5 (or why this wouldn't be outsourced to an Electronic Resource Management system), but that didn't seem to get much uptake in our conversation.

I can definitely check in on what the interpretation of "double clicks" of versioned items is from a COUNTER perspective.

**Comment by NateWr** [comment]
To my surprise, I think we are ok-ish on the versioning front without any changes. However, there is one possible issue I uncovered while looking into it.

From my investigations, here's how the URLs are parsed:

- Anything with `article/view` is considered `ASSOC_TYPE_SUBMISSION`.
- Anything with `article/download` is considered `ASSOC_TYPE_SUBMISSION_FILE`.

This works for PDF/HTML galleys because even though they load at `article/view/<submission-id>/<galley-id>`, the PDF or HTML file is loaded in an iframe at `article/download/<submission-id>/<galley-id>`.

Our versioned URLs look like `article/view/<submission-id>/version/<publication-id>[/<galley-id>]`, but the underlying PDF or HTML file is still loaded in `article/download/<submission-id>/<galley-id>`. So visits to the landing page and galleys of old versions are counted correctly.

However, this means that when someone directly visits the page of a PDF or HTML galley, without going through the article landing page, they record two entries in the logs: `article/view/<submission-id>/<galley-id>` and `article/download/<submission-id>/<galley-id>`.

`UsageStatsLoader::_getUrlMatches()` returns both lines as a match, one for submission and one for the file (what we call in the backend stats area an "abstract" hit and a "files" hit). However, once the file is parsed, it appears in the `usage_stats_temporary_records` table as a single hit to the submission (not the file):

```
mysql> select * from usage_stats_temporary_records where day="20191219";
+----------+------------+----------+------------+--------+------------+--------+------+---------------------------+-----------+
| assoc_id | assoc_type | day      | entry_time | metric | country_id | region | city | load_id                   | file_type |
+----------+------------+----------+------------+--------+------------+--------+------+---------------------------+-----------+
|        5 |    1048585 | 20191219 | 1576762274 |      1 | NULL       | NULL   | NULL | usage_events_20191219.log |         0 |
+----------+------------+----------+------------+--------+------------+--------+------+---------------------------+-----------+
```

This is not a new issue with versioning, but I thought I would double-check with you @asmecher and @ctgraham to see if this is intended behaviour. It seems to me like this should count as a single view of the file, not the submission.

To test this I constructed the following fake log to parse, which includes only two URL hits. One to the URL to view a PDF (`article/view/5/1`) and one to the URL that actually loads the PDF (`article/download/5/1/15)`. I believe that this simulates what a single request to view a PDF would generate in a real usage log.

```
127.0.0.1 administrative 1 "2019-12-19 14:32:08" http://localhost:8000/publicknowledge/article/view/5/1 200 "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:71.0) Gecko/20100101 Firefox/71.0"
127.0.0.1 administrative 1 "2019-12-19 14:32:08" http://localhost:8000/publicknowledge/article/download/5/1/15 200 "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:71.0) Gecko/20100101 Firefox/71.0"
```

(To use this for your own testing you would need to update the submission, galley and file ids to ones that exist in your system.)

**Comment by asmecher** [comment]
I think this is probably tied up with COUNTER, which I'm not very familiar with -- it has specifications for things like debouncing. And I'm not sure whether COUNTER business rules are applied when metrics are recorded, or when they're processed. @ctgraham, are you familiar with this aspect? If not, maybe I could follow up with Bozana.

**Comment by ctgraham** [comment]
I don't think COUNTER cares about abstract views, nor has it in my memory.  I reviewed COUNTER R5, R4, and R3 and each focuses on fulltext downloads/views.  Back in R3 there was a distinction between fulltext HTML, fulltext PDF, and fulltext "other", but that distinction goes away in later releases.

If a user views fulltext via HTML and then fulltext via PDF, or downloads the fulltext via the same medium twice, this only counts a one view.

If I were more helpful I would check to make sure the COUNTER reports are using only ASSOC_TYPE_SUBMISSION_FILE for calculations (I think this is the case), check for legacy reports in OJS which describe abstract views separately from fulltext views (hopefully not), and would verify that the inline display of HTML fulltext is registered correctly (I think I recall fancy jiggering in the a plugin for this).  But first I want to get that Crossref testing out the door, and even that simple task is eluding me right now.

**Comment by NateWr** [comment]
Thanks @ctgraham. I think this is not an urgent question to address for 3.2, because is not something newly introduced. I'll defer it from 3.2 for now but would like to keep this conversation open.

I'm not sure if I understand you correctly, but perhaps there is some divergence here between how OJS keeps statistics and how you describe COUNTER expectations.

First, OJS does make a distinction between abstract and file views. It tracks both separately in the metrics type by the `assoc_type` and we display them independently to editors in the new article stats UI.

Second, from my brief test, it appears that OJS is treating a single visit to the view PDF page as a single visit to the article abstract page. If COUNTER ignores abstract views entirely, usage may be under-reported. (This was an isolated test. More investigation would be necessary to see what happens in different scenarios and how the rows in a metrics table get compiled into COUNTER reports.)

**Comment by bozana** [comment]
Maybe I can take a further look at the problems described here i.e. what would be needed to solve for the next release...
First I will test a few things to get the current status. I will do it incrementally because I can only work a little bit every day, will write the (part) results here, and then at the end come back to you all. Thus, please ignore the text till my testing is finished.

OJS:
1. Galley view and download:
"However, this means that when someone directly visits the page of a PDF or HTML galley, without going through the article landing page, they record two entries in the logs: article/view/<submission-id>/<galley-id> and article/download/<submission-id>/<galley-id>"
When going to a galley directly, e.g. `http://.../index.php/publicknowledge/article/view/1/1` I only see one log entry: `... http://.../index.php/publicknowledge/article/download/1/1/2 200 ...`
i.e. the galley landing page 'http://.../index.php/publicknowledge/article/view/1/1' is never logged.
This conforms to the code in the UsageEventPlugin: https://github.com/pkp/ojs/blob/master/plugins/generic/usageEvent/UsageEventPlugin.inc.php#L57-L71. 
However, journal might use apache logs where the galley view page is logged, so I will also test how the log file is processed. It should actually follow the same logic, but...

2. Abstract pages and downloads:
We do log both, i.e.: journal landing page, issue landing page, article landing/abstract page, as well as file downloads (article and issue galleys). 
How are they considered in the current report tools? 

3. Considering versioning:
a) I have no idea if there is a COUNTER rule for visiting different article versions but counting it only once -- it feels like they should be counted separately, but... I will try to find something about it...
b) The 'version' URL paths seem not to be considered, e.g. article abstract page of an article old version (http://.../index.php/publicknowledge/article/view/2/version/3) seems not to be logged.

**Comment by bozana** [comment]
Hi @NateWr and @ctgraham, there would be some things to correct here: the problem that Nate found out (#1 and #3 above) should be fixed, but also some changes in the the log processing (#4b above) have to be made. May I ask you to read that #4 and related COUNTER link/document above? -- to be sure that I understand it correctly. 
Also, I could not find anything specific about different article versions -- if they are considered as a unique item. I could try to write to that email address on the counter page, if you have no better idea. Or should/can we maybe just define them as one item for now?
I will take a further look into the Release 5 of the COUNTER reports, and then come back to Clinton again. Or maybe you already know what is to do and if we should update?
Thanks a lot!!!

**Comment by bozana** [comment]
Hi @NateWr and @ctgraham, reading the document https://www.projectcounter.org/wp-content/uploads/2020/08/Module_2_Journal_Usage_20200811.pdf: it seems that that unique item and title (from the 4b and 4c above) first came now with the release 5. The release 4, that we currently support, I think, still counts HTML and PDF separately. Thus, I would maybe suggest that we only fix the problems Nate encountered for this issue (eventually also OJS/OMP/OPS release?) and then think and re-implement the statistics so that we support that release 5. What do you think?

**Comment by bozana** [comment]
Hmmm... This Release 5 seems to has been out there for almost 3 years now, so I suppose we should support it very soon...

**Comment by ctgraham** [comment]
COUNTER R4 is still fairly widely used around Libraries, though it ought to be phased out in favor of R5; Pitt ULS uses COUNTER in communicating usage statistics to Plum Analytics.

@shanu17 from Pitt ULS is working on SUSHI/COUNTER R5 for PKP.  I owe him better definitions of each report so that he can map the report requirements against our Statistics Service / MetricsDAO.  There remain some gaps in our internal statistics harvesting for non-OA usages, e.g. mapping access against institutional subscription and counting access denied requests.

**Comment by bozana** [comment]
Hi @ctgraham and @shanu17, that is great to hear -- that you are working on the support for R5! :-))) 
I would then only fix the processing problems @NateWr encountered here, for R4, now. 
If I can somehow help for the support of R5, just let me know. (Maybe we can open a new issue for that and ... )

**Comment by bozana** [comment]
The coming PRs consider the following issues:

1) Article abstract page versions URL were not logged. Now they are logged, e.g. `.../article/view/1/version/1`.

2) If there is a log entry of a galley view page, it will not be counted.
The log file example above:
```
127.0.0.1 administrative 1 "2019-12-19 14:32:08" http://localhost:8000/publicknowledge/article/view/5/1 200 "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:71.0) Gecko/20100101 Firefox/71.0"
127.0.0.1 administrative 1 "2019-12-19 14:32:08" http://localhost:8000/publicknowledge/article/download/5/1/15 200 "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:71.0) Gecko/20100101 Firefox/71.0"
```
counts only the last, file download.

3) The representation ID was calculated using `file->getAssocId()`. Because a file can be associated with several representations in OJS now, this eventually led to saving the wrong representation ID in the table metrics. This is corrected -- now the representation ID is passed through, from the log file URL processing via table `usage_stats_temporary_records` to the table `metrics`.

NOTE about the double click processing for versions:
The double clicks works with assocType and assocId -- if they are equal, the double click rule is applied. That means for versioning:
For the file download, e.g. the following log file entries when two different versions have the same file:
```
article/download/2/3/4
article/download/2/4/4
```
are counted only 1.

If file changes in a new version, e.g. the following log file entries
```
article/download/1/1/2
article/download/1/10/18
```
are counted = 2.

This seems to be OK -- if the file does not change and the new version contains the same file (with the same file ID) it is considered for double click.
But if the abstract changes (which we do not know) and we have these two URLs in the log file:
```
article/view/1/version/1
article/view/1/
```
they are counted = 1.

Can this stay so? Else, we would need to consider publication ID as well.
(I believe the double clicks are mostly meant to deal with the same URLs in R5, so we will eventually change the current implementation of the double click processing here https://github.com/pkp/pkp-lib/issues/6781)

**Comment by bozana** [comment]
PRs:
pkp-lib: https://github.com/pkp/pkp-lib/pull/6787
usageStats: https://github.com/pkp/usageStats/pull/43
ojs: https://github.com/pkp/ojs/pull/3051

**Comment by bozana** [comment]
@NateWr, could you please take a look at the PRs i.e. changes above? -- I haven't tested them with OMP and OPS, so just a rough view would be sufficient -- if the solutions are OK so -- and I would continue with OMP and OPS -- and then later a deeper code review can happen. Please see my comment above -- what they are solving now.
Also, do you know what is to do so that the new db table structure (new table column) is ensured for the users after these changes are merged -- I've forgotten it... :-\
Thanks a lot!

**Comment by NateWr** [comment]
Thanks @bozana! I've had a quick look and it looks ok to me. 

> Thus, I am not sure if this applies to the same URLs or content objects.

You've probably assessed the COUNTER spec closer than me, but my sense is that we should treat versions as equivalent. So a click on any version of an article will count towards that article and we don't need to track versions separately.

This is a little bit more complicated for galleys, because we track clicks to galleys independently. And there is no way to say Galleys A and B are two versions of the same Galley. But I'm not sure this matters for us, because to my knowledge we do not present statistics on clicks to individual galleys. We only use the representation tracking to determine whether it was a click on a PDF, HTML or Other representation.

Maybe in the future we would want to display the usage stats for every representation. But I can't recall any requests for that. This probably deserves a broader discussion. But for the purposes of this PR, I think (I hope!) that it will be enough to record visits to galleys as counting towards a hit on a submission, regardless of what version the galley is. In other words, I think we can disregard the versioning in the stats.

**Comment by bozana** [comment]
Thanks a lot @NateWr!
I will then test everything with OMP and OPS and make new PRs...

Maybe here just one general note:
All the abstract views, as well as all files downloads of a submission, no matter which version is counted towards that article in total. For a submission we then usually display the numbers for abstracts, PDF downloads, HTML downloads, eventually other downloads, and a total count, so the versions do not really matter. We however have the information about the file and representation ID, so that we could differentiate between the files and even versions for file downloads, which is not the case for abstract views (which is OK, I think, and as you said).
When we implement the R5 it will be similar, but we would additionally need to keep the information about the unique item numbers.

I am currently not sure about our double click processing -- as described above -- but I think we can leave that for now as it is. In R5 they are defined more precise, only the same URLs should be considered.
If two different abstract versions (that have different URLs) are accessed within 10 seconds, they would now count only 1. In R5 they would count total = 2, unique item = 1.
Thus, because IMO the R4 does not define it very precise, and because only that one or two cases regarding our versioning (that are probably also very rare to happen) is affected by the rule, we can leave it so.

**Comment by bozana** [comment]
@NateWr, shall all the changes be ported to the stable-3_3_0 even if the DB upgrade is needed?

**Comment by NateWr** [comment]
I don't think a 3.3.0-x release can have a migration in it. Are we sure that this is needed? I know the problem is that a single file id might track back to multiple representation ids. But if all we need to know is whether it is PDF, HTML or Other, can we get away with not knowing the representation ID? Can we get that from the file ID / path itself?

**Comment by bozana** [comment]
The DB table metrics already has the column representation_id and currently it could contain the wrong IDs there. It seems that that column is currently not really used -- it was used earlier for the submission usage stats (graph) represented to the reader on the submission view page -- probably to display the usage of each galley/publication format (OMP). Those usage (graph) display functions seem not to be used any more. 
If we only want to differentiate between PDF, HTML and Other -- no matter about galley, publication/version or publication format (OMP) -- we could ignore that column i.e. remove it later. But, in general it is maybe OK/useful to have that information too.
Unfortunately I do not see any other way to get the right galley ID (in OJS) when we only know the file ID.
Thus, maybe we can port the other changes to the stable branch -- that do not need the DB upgrade -- and have that change with the DB upgrade first in the next major release 3.3.1?

**Comment by NateWr** [comment]
Yeah, I think we need to do what we can without a DB upgrade for now. We can do the DB upgrade with 3.3.1, but I think in general that we have discussed reducing the metrics table columns where possible. I can see the utility of collecting whatever we can, but I also know that some large installations have struggled with the sheer size of their metrics table. Let's discuss this further as we think about how to refactor our usage stats.

**Comment by bozana** [comment]
@NateWr, could you take another look at the PRs above? -- Now everything is finished for the main branch...
I will back port those fixes not needing the DB upgrade to the stable-3_3_0...
Thanks a lot!

**Comment by NateWr** [comment]
:+1:  I've done a code review but not yet tested it myself. When you have PRs for `stable-3_3_0` I can load it up and run it through some manual tests.

**Comment by bozana** [comment]
Hi @NateWr, sorry to bother you again -- I've considered everything from the code review I think -- if you would like to have another look. Thanks a lot!

**Comment by NateWr** [comment]
Looks good to me. :+1:

**Comment by bozana** [comment]
Thanks a lot @NateWr! I would then merge the main branch changes as soon as all tests are successfully run, ok?

I also did the PRs for stable-3_3_0 (see above). Those contain only the fixes for the problem No. 1 and No. 2 from here https://github.com/pkp/pkp-lib/issues/4904#issuecomment-784346753. (No 3. requires DB change, so this is not coming into the stable branch). Also, I did not implement PKPSubmissionDAO::exists() -- as we said -- not to change it in the stable branch, but I added the PKPPublicationDAO::exists() -- because this check is new/first now added.
Would you like to take a look and eventually test it?
:pray:

**Comment by NateWr** [comment]
Looks good, go ahead and merge to stable.

**Comment by bozana** [comment]
Everything merged from this issue, thus closing...
